presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java

```diff
@@ -2496,14 +2502,82 @@ private AccumulatorFactory buildAccumulatorFactory(
+            InternalAggregationFunction internalAggregationFunction = metadata
+                    .getFunctionRegistry()
+                    .getAggregateFunctionImplementation(aggregation.getSignature());
+
+            List<Integer> valueChannels = new ArrayList<>();
             for (Expression argument : aggregation.getCall().getArguments()) {
-                if (argument instanceof LambdaExpression) {
-                    // Wenlei Hack TODO: Compile lambda
-                    continue;
+                if (!(argument instanceof LambdaExpression)) {
+                    Symbol argumentSymbol = Symbol.from(argument);
+                    valueChannels.add(source.getLayout().get(argumentSymbol));
+                }
+            }
+
+            List<LambdaChannelProvider> lambdaChannelProviders = new ArrayList<>();
+            List<LambdaExpression> lambdaExpressions = aggregation.getCall().getArguments().stream()
+                    .filter(LambdaExpression.class::isInstance)
+                    .map(LambdaExpression.class::cast)
+                    .collect(toImmutableList());
+            if (!lambdaExpressions.isEmpty()) {
+                List<FunctionType> functionTypes = aggregation.getSignature().getArgumentTypes().stream()
+                        .filter(typeSignature -> typeSignature.getBase().equals(FunctionType.NAME))
+                        .map(typeSignature -> (FunctionType) (metadata.getTypeManager().getType(typeSignature)))
+                        .collect(toImmutableList());
+                List<Class> lambdaInterfaces = internalAggregationFunction.getLambdaChannelInterfaces();
+                verify(lambdaExpressions.size() == functionTypes.size());
+                verify(lambdaExpressions.size() == lambdaInterfaces.size());
+
+                for (int i = 0; i < lambdaExpressions.size(); i++) {
+                    LambdaExpression lambdaExpression = lambdaExpressions.get(i);
+                    FunctionType functionType = functionTypes.get(i);
+
+                    // To compile lambda, LambdaDefinitionExpression needs to be generated from LambdaExpression,
+                    // which requires the types of all sub-expressions.
+                    //
+                    // In project and filter expression compilation, ExpressionAnalyzer.getExpressionTypesFromInput
+                    // is used to generate the types of all sub-expressions. (see visitScanFilterAndProject and visitFilter)
+                    //
+                    // This does not work here since the function call representation in final aggregation node
+                    // is currently a hack: it takes intermediate type as input, and may not be a valid
+                    // function call in Presto.
+                    //
+                    // TODO: Once the intermediate representation hack is fixed, the same mechanism
+                    // in project and filter expression should be used here.
```

Comment:
I think the approach you took here is pretty reasonable. I don't consider it a hack.

Reply:
The hack is for aggregation function's intermediate representation. i.e. how the final aggregation is represented today .